### ISSUE TITLE
visual extractor

### ISSUE THREAD

**Issue by Justarrrrr** [GaryGuTC/COMG_model#21]
Hello, thank you very much for your excellent work on COMG and ORID, and I would like to ask here about the difference in the extraction methods of the two models for organ-level fine-grained features.
First ,this is the code  of  COMG  for  catch fine-grained image information: !(https://github.com/GaryGuTC/COMG_model/blob/main/COMG_model/modules/visual_extractor.py#L158)
![image](https://github.com/user-attachments/assets/9d1db1d5-18ca-4d15-8154-b5c371bd39da)
It seems that the mask image features are multiplied by the original image features and then obtained by ResNet18, which is consistent with the graph in your COMG paper:
![image](https://github.com/user-attachments/assets/33c45e03-44c3-445d-91ce-3c0bac953238)
But according to your ORID diagram:
![image](https://github.com/user-attachments/assets/cbbdf78b-f2c3-4aba-906d-96eae6eb6df5)
It seems to be extracted by ResNet first and then multiplied.
I would like to ask if there is a difference in the effectiveness of the two methods.

**Comment by GaryGuTC** [comment]
Hi
Thanks so much for your attention to our works.
So sorry for the difficulties caused by our unclear expression.
I have provided the code of the visual extractor in the ORID model.
I hope it will be helpful to you.

```python
class VisualExtractor(nn.Module):
    def __init__(self, args):
        super(VisualExtractor, self).__init__()
        
        self.visual_extractor = args.visual_extractor
        self.pretrained = args.visual_extractor_pretrained
        
        self.model = Reconstructed_resnet101(self.pretrained)
        self.avg_fnt = torch.nn.AvgPool2d(kernel_size=7, stride=1, padding=0)
        
        self.model18 = Reconstructed_resnet18(self.pretrained)
        self.model_bone = Transfer_model(args.mask_bone_num)
        self.model_lung = Transfer_model(args.mask_lung_num)
        self.model_heart = Transfer_model(args.mask_heart_num)
        self.model_mediastinum = Transfer_model(args.mask_media_num)
        self.model_pleura = Transfer_model(args.mask_pleural_num)
        self.model_list = [
            self.model_bone,
            self.model_lung,
            self.model_heart,
            self.model_mediastinum,
            self.model_pleura
        ]

    def forward(self, 
                images,
                pleural_mask,  
                lung_mask, 
                mediastinum_mask, 
                heart_mask,  
                bone_mask): 
        bs, mask_img_h, mask_img_w = images.size(0), pleural_mask.size(2), pleural_mask.size(3)
        
        tmp_feature, patch_feats = self.model(images)  
        avg_feats = self.avg_fnt(patch_feats).squeeze().reshape(-1, patch_feats.size(1))
        
        batch_size, feat_size, _, _ = patch_feats.shape
        patch_feats = patch_feats.reshape(batch_size, feat_size, -1).permute(0, 2, 1) 
        
        img_local_feature = tmp_feature.reshape(bs, 1, mask_img_h, mask_img_w) 
        
        organ_feature_list = []
        for (organ_mask, organ_module) in zip([bone_mask, lung_mask, heart_mask, mediastinum_mask, pleural_mask], self.model_list):
            organ_meaning_feat = self.model18(
                organ_module(
                    img_local_feature.expand(bs, organ_mask.size(1), mask_img_h, mask_img_w) * organ_mask
                ) 
            ) 
            organ_meaning_feat = organ_meaning_feat.reshape(bs,organ_meaning_feat.size(1),-1).permute(0,2,1) 
            organ_feature_list.append(organ_meaning_feat)
        
        return patch_feats, avg_feats, organ_feature_list
```

Best regards,
Tiancheng

**Comment by Justarrrrr** [comment]
Thank you very much for your reply, then I have one more question, are the args here are int:
![image](https://github.com/user-attachments/assets/18cdaa13-7da0-4afa-824d-d4c8b93e271f)
The number of channels that represent the mask feature：
![image](https://github.com/user-attachments/assets/3357af5a-13b7-42c9-9e88-bb18a7e51abb)

**Comment by GaryGuTC** [comment]
That's correct, The same as [here](https://github.com/GaryGuTC/COMG_model/blob/467bc7d7b517f9183063355940c6f33db521d33b/COMG_model/modules/visual_extractor.py#L50C1-L61C46).
